doc/hwlookup.1
.ad b
.TH hwlookup 1 "23 July 1999" "CORE software" "CORE software"
.AT 3
.de sh
.br
.ne 5
.PP
\fB\\$1\fR
.PP
..
.PP
.SH NAME
.PP
.I hwlookup
- query the ethernet hardware address database off-line.
.PP
.SH SYNOPSIS
.PP
.I hwlookup
[options] address [ address [ ... ] ]
.PP
.SH DESCRIPTION
.PP
when 
.BR hwpassive (8)
exits, it leaves behind a database of the new
address mappings it has found.  while 
.BR hwpassive (8)
is running, you
are incouraged to use its set of 
.BR dcp (1)
commands to query the
database realtime.  when it is not running, you can use 
.I hwlookup
instead.
.PP
addess can either be an IP address, or a ethernet hardware address.  IP
addresses must be specified as a dotted decimal, but may be followed by a
netmask.  ie.
.PP
% hwlookup 150.135.0.0/16
.PP
returns all entries for the 150.135 subnet.  for ethernet hardware
addresses, you need only specify the first digits if you wish, ie
.PP
% hwlookup 00:d0:c0
.PP
will match all ethernet hardware addresses beginning with the prefix
00:d0:c0.
.PP
there are two output modes, short and long.  the long mode is default and
spits out more information.  here is an example of a long output
.PP
==============================================
HW:             00:c0:4f:93:6d:8c
IP:             128.196.51.111
first seen:     Fri Jul 23 11:42:32 1999
last seen:      Fri Jul 23 12:18:34 1999
card info:      Dell
.PP
.B HW
and 
.B IP
are clearly the hardware/IP address mappings.  
.B first seen
and 
.B last seen
are respectively the first and last time
the mapping had been seen by 
.BR hwpassive (8).
.B card info
specifies the
vendor who uses the 3 byte prefix, if found in the hwcode database of
known manufacturer/prefix mappings.
.PP
note that the database file used by hwpassive and hwlookup are specific
to their version of netl, and are also not platform portable.
.PP
.SH OPTIONS
.PP
-v, -r, --resolve, --dont-resolve, -d, --debug, --help work as described
in the 
.BR netl (8)
man page.
.PP
.PP
.B -f
.I filename
, 
.B --file
.I filename
	specifies the database file which should be read, instead of
/usr/local/lib/netl/hwpassive.
.PP
.B -s
, 
.B --short
	use short output mode.  only the hardware/IP address mapping will
be printed on a single line.
.PP
.B -l
, 
.B --long
	use long output mode.  the address mapping will be followed by
the time the mapping was first seen, last seen, and the manufacturer of
the card, if known.
.PP
.SH HISTORY
.PP
the genesis for this program came about when the author and a friend of
his were tracking quake users on their university LAN.  thanks to the
networking knowledge of said friend, determined programming of the author,
and dominos pizza, we were able to track nearly every anonymous quaker to
an email address, phone number, dorm, and sometimes even their room
number.
.PP
note that neither the author, nor the author's friend play quake anymore,
and do not condone the senseless violence in many of today's video games.
.PP
.SH SEE ALSO
.PP
.BR netl (8),
.BR netl.conf (5),
.BR netlcc (1),
.BR netl_install (1),
.BR netl_module (1),
.BR neta (1),
.BR hwpassive (8),
.BR hwlookup (1),
.BR dcp (1)
and 
.BR xd (1)
.PP
.SH BUGS
.PP
there are almost certainly bugs, please report them to me.  send bugs and
bug fixes to netl@netl.org.  the netl home page is at
http://www.netl.org, which should contain up to date information on
.B netl
.
.PP
i have attempted to write pretty readable documentation, however, i'm not
really the best technically writer.  if you are, maybe we could
colaborate?
.PP
.SH COPYING
.PP
Copyright 1996, 1997, 1999 Graham THE Ollis
.PP
This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 2 of the License, or (at your
option) any later version.
.PP
This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
Public License for more details.
.PP
You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
675 Mass Ave, Cambridge, MA 02139, USA.
.PP
.PP
.SH AUTHOR
.PP
Graham THE Ollis <ollisg@netl.org>


HISTORY
version .95.2
-------------

 . mostly internal stuff.  added parse.c and grab.c modules, which contain
   code that used to be in the netl.c main module.  this is so that all the 
   network specific code is in grab.c, the true nightmare for porting is
   in this file... so if i can port grab.c to win32... then i'm done.

 . the \ character now works in the config file, if you want to have
   multiple line rules.  very unix-esq.

version .95
-----------

 . created man page documentation.

 . added `detect' config line, see the netl(8) manpage for details.
   this will reduce the amount of configeration you will have to do
   in the netl.conf file.

 . modified dist version of netl.conf so that it's more portable.
   in particular, i put in literal port numbers rather than using
   port names.

 . dcp.pl uses perl native Socket module instead of using net cat.
   `dcp.pl -o' uses the old method.

version .94
-----------

 . changed all the struct declerations to typedef declerations.  get off my
   back ok?  i've been programming perl for so long, forgot some of that
   strange C language. 

 . no longer depending on linux .h files for IP (and sub) headers.  this
   is for portability reasons.  if you are doing a port, make sure you get
   the endian type right!  there are now three global headers:
	global.h - general global stuff (endian integer conversion etc.)
	ether.h - ethernet specific data types, #defines
	ip.h - ip protocol specific data types, #defines
   logically, if your adding an interface (ATM, token ring, ppp, etc.)
   you'd want to start in ether.h.  if your adding an OS, look at global.h
   first.  if you want to add a protocol (IPX, etc.) start with ip.h and
   modify/add to these files.  if your trying to do all of this at once,
   god help you. 

 . fixed the clearipcache() function.  thus now, when you send a 
   netl:readconfig, netl will clear the ip cache before reading the file,
   avoiding all sorts of potential duplicates.  though you probably have
   an enourmous amount of memory like me and don't care... 

 . added some global options with ports to win32 in mind.  NO_SYSLOGD, if 
   defined in global.h will not even compile with the syslogd library stuff.  
   this is a good idea if your running from a non unix system.  note that 
   syslogd is really nice in that it partially compresses lines by indicating 
   repeated lines.  this won't save you if you get strobed, but will keep your 
   log file neat if some one ping -f's you from only one site.

 . added NO_TEEOUT option which is defined for unix.  the -o option should 
   only be necessary in win32 only.  even then, you should only compile with
   this option if your thinking of using it.  -o copies output to the specified
   file.

0.93 (and 0.92)
---------------

 a. -z option allows you to run netl in the foreground.  output is sent to
    stdout instead of syslogd.

 b. you can include configeration lines in the command line.  when you do
    this, the config file /etc/netl.conf will NOT be read.  note that (a)
    and (b) are meant to be used in concert with each other for testing
    and quick monitoring. 

 c. dcp is partially implemented.  netl will listen to port 47 udp
    (silently) if you `listen 47' in your config file.  when you send the
    message `netl:readconfig' netl is suppose to re read your config file. 
    dcp.pl is a perl script (you'll need net cat for this to work) which
    sends this message.  what is left to do for dcp?
	1. generic server side dcp interface libdcp.a file or somesuch.
	2. more solid C version of dcp client.

 d. netl is now officially under the GPL, see COPYING for more information.

 e. icmp codes are now included in some .h file somewhere rather than
    depending on your linux being overly up to date.

 f. neta no longer swaps source/destination ip/ethernet addresses
    inapropriately. 

 g. started work on hwpassive, which is no where near complete.

 h. tdr.pl is no included with netl distribution for convenience.  get
    your own version at: 

http://www.u.arizona.edu/~ollisg/core.html
	[ CORE ]

netl 0.91
---------

 minor updates

netl 0.90
---------

 initial release


hwlookup.c
/*==============================================================================
| hwlookup
|   optimized (and debugged) by Graham THE Ollis <ollisg@wwa.com>
|
|   Copyright (C) 1997 Graham THE Ollis <ollisg@wwa.com>
|
|   This program is free software; you can redistribute it and/or modify
|   it under the terms of the GNU General Public License as published by
|   the Free Software Foundation; either version 2 of the License, or
|   (at your option) any later version.
|
|   This program is distributed in the hope that it will be useful,
|   but WITHOUT ANY WARRANTY; without even the implied warranty of
|   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
|   GNU General Public License for more details.
|
|   You should have received a copy of the GNU General Public License
|   along with this program; if not, write to the Free Software
|   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
|
|  Date       Name	Revision
|  ---------  --------  --------
|  23 Jul 99  G. Ollis	wrote
|=============================================================================*/

static char	*id = "@(#)hwlookup by graham the ollis <ollisg@wwa.com>";
static void dumb(char *d) { dumb(id); }	/* this avoids an anoying error msg */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <time.h>

#include "netl/global.h"
#include "netl/options.h"
#include "netl/hwpassive.h"
#include "netl/resolve.h"
#include "netl/io.h"

typedef struct address_Entry {
	u8 addr[6];
	u8 bitmask[6];
	int flag;	/* TRUE if we are meant to NOT this entry */
	struct address_Entry *next;
} address_entry;

static address_entry *hws = NULL, *ips = NULL;
static int count = 0;

static int modifyip(u32 *num, char *name);
static int modifyhw(u8 *answer, u8* mask, char *name);
static void printentry(hwpassive_entry *e);

/*==============================================================================
| int main(int, char **)
|=============================================================================*/

int
main(int argc, char *argv[])
{
	FILE *fp;
	char *mask_s = NULL;
	int mask;
	int i, n;
	hwpassive_entry e;

	prog = argv[0];
	configfile = NETL_LIB_PATH "/hwpassive";

	//setservent(TRUE);
	parsecmdline(argc, argv); 
	if(displayVersion) {
		fputs("hwlookup ", stdout);
		puts(COPYVER);
	}

	for(i=1, n=0; i<argc; i++) {
		if(argv[i][0] != '-') {	/* it's not an option */
			char *ptr = argv[i];
			int flag = FALSE;

			mask = -1;
			mask_s = strrchr(ptr, '/');
			if(mask_s != NULL) {
				mask_s[0] = 0;
				mask_s++;
				mask = atoi(mask_s);
			}
			if(ptr[0] == '!') {
				flag = TRUE;
				ptr++;
			}

			if(!strcmp("all", ptr)) {
				u32 addr = 0;
				u32 addr_mask = 0;
				address_entry *tmp;
				tmp = allocate(sizeof(address_entry));
				memcpy((void *) tmp->addr, &addr, sizeof(u32));
				memcpy((void *) tmp->bitmask, &addr_mask, sizeof(u32));
				tmp->next = ips;
				ips = tmp;
				n++;
			} else if(strchr(ptr, ':') != NULL) { /* it's a hw address */
				u8 addr[6];
				u8 addr_mask[6] = {	0xff, 0xff, 0xff,
							0xff, 0xff, 0xff  };
				if(mask == -1)
					mask = 48;
				if(modifyhw(addr, addr_mask, ptr)) {
					address_entry *tmp = allocate(sizeof(address_entry));
					memcpy(tmp->addr, addr, 6);
					memcpy(tmp->bitmask, addr_mask, 6);
					tmp->flag = flag;
					tmp->next = hws;
					hws = tmp;
					n++;
				}


			} else {		/* otherwise, it's IP */
				u32 addr;
				u32 addr_mask = 0xffffffff;
				if(mask == -1)
					mask = 32;
				addr_mask = htonl(addr_mask << (32 - mask));
				if(mask == 0)
					addr_mask = 0;
				if(modifyip(&addr, ptr)) {
					address_entry *tmp;
					tmp = allocate(sizeof(address_entry));
					memcpy((void *) tmp->addr, &addr, sizeof(u32));
					memcpy((void *) tmp->bitmask, &addr_mask, sizeof(u32));
					tmp->flag = flag;
					tmp->next = ips;
					ips = tmp;
					n++;
				}

			}
		}
	}

	if(n == 0) {
		fprintf(stderr, "%s: no search criteria given, die\n", prog);
		return 0;
	}

	if(debug_mode) {
		while(ips != NULL) {
			printf("ip:%s/%08x\n", 
				ip2string(*((u32 *) ips->addr)),
				*((u32 *) ips->bitmask));
			ips = ips->next;
		}
		while(hws != NULL) {
			printf("hw:%02x:%02x:%02x:%02x:%02x:%02x",
				hws->addr[0], hws->addr[1], hws->addr[2],
				hws->addr[3], hws->addr[4], hws->addr[5]);
			printf("/%02x:%02x:%02x:%02x:%02x:%02x\n",
				hws->bitmask[0], hws->bitmask[1], hws->bitmask[2],
				hws->bitmask[3], hws->bitmask[4], hws->bitmask[5]);
			hws = hws->next;
		}
		return 0;
	}

	fp = fopen(configfile, "r");
	if(fp == NULL) {
		fprintf(stderr, "%s: error opening %s for read\n", prog, configfile);
		return 1;
	}

	while(fread(&e, sizeof(hwpassive_entry), 1, fp) == 1) {
		address_entry *tmp;
		int done = FALSE;
		u8 ip[4];
		memcpy(ip, &e.ip, 4);

		tmp = ips;
		while(tmp != NULL && !done) {
			int i, b=TRUE;
			//printf("************************\n");
			for(i=0; i<4; i++) {
				//printf("ip[%d] = %d\n", i, ip[i]);
				//printf("tmp->addr[%d] = %d\n", i, tmp->addr[i]);
				//printf("tmp->bitmask[%d] = %d\n", i, tmp->bitmask[i]);
				if(!	((ip[i] & tmp->bitmask[i]) ==
					 (tmp->addr[i] & tmp->bitmask[i]))
				  ) {
					b = FALSE;
				}
			}
			//printf("b = %d\n", b);
			//printf("tmp->falg = %d\n", tmp->flag);
			//printf("b ^ tmp->flag = %d\n", b ^ tmp->flag);
			//printf("************************\n");

			if(b ^ tmp->flag) {
				printentry(&e);
				done = TRUE;
			}
			tmp = tmp->next;
		}

		tmp = hws;
		while(tmp != NULL && !done) {
			int i, b=TRUE;
			for(i=0; i<6; i++) {
				if(!	((e.hw[i] & tmp->bitmask[i]) ==
					tmp->addr[i])
				  ) {
					b = FALSE;
				}
			}
			if(b) {
				printentry(&e);
				done = TRUE;
			}
			tmp = tmp->next;
		}
	}

	fclose(fp);

	if(hwlookup_mode == HWLOOKUP_COUNT)
		printf("%d\n", count);

	return 0;
}

static char *
time2str(time_t t)
{
        static char buffer[40];
        strcpy(buffer, ctime(&t));
        *strchr(buffer, '\n') = 0;
        return buffer;
}

static char *
time2str2(time_t t)
{
        static char buffer[40];
        strcpy(buffer, ctime(&t));
        *strchr(buffer, '\n') = 0;
        return buffer;
}

static char *
cardinfo(u8 *hw)
{
	FILE *fp;
	static char buffer[1026];
	char *num, *text=NULL;
	char tmp[3];

	fp = fopen(NETL_LIB_PATH "/hwcode", "r");
	if(fp == NULL)
		return "hwcode missing!";

	while(fgets(buffer, 1024, fp) != NULL) {
		int b = TRUE, n=0;
		num = strtok(buffer, ":\n");
		if(num != NULL && (text = strtok(NULL, ":\n")) != NULL) {
			while(num[0] != 0 && b) {
				u8 data;
				tmp[0] = num[0];
				tmp[1] = num[1];
				tmp[2] = 0;
				data = ahextoi(tmp);
				if(data != hw[n++])
					b= FALSE;
				num++;
				if(num[0] != 0)
					num++;
			}
			if(b)
				return text;
		}
	}
	return "unknown";
}

static void
printentry(hwpassive_entry *e)
{
		count++;
	if(hwlookup_mode == HWLOOKUP_SHORT) {
		printf("%02x:%02x:%02x:%02x:%02x:%02x => %s\n",
			e->hw[0], e->hw[1], e->hw[2], 
			e->hw[3], e->hw[4], e->hw[5],
			ip2string(e->ip));
	} else if(hwlookup_mode == HWLOOKUP_COUNT) {
	} else {
		printf(	"==============================================\n" 
			"HW:\t\t%02x:%02x:%02x:%02x:%02x:%02x\n"
			"IP:\t\t%s\n"
			"first seen:\t%s\n"
			"last seen:\t%s\n"
			"card info:\t%s\n",
			e->hw[0], e->hw[1], e->hw[2], 
			e->hw[3], e->hw[4], e->hw[5],
			ip2string(e->ip),
			time2str(e->first), time2str2(e->last),
			cardinfo(e->hw));
	}
}

static int
modifyhw(u8 *answer, u8 *mask, char *name)
{
	char *e;
	int i = 0;

	for(i=0; i<6; i++) {
		answer[i] = 0;
		mask[i] = 0;
	}

	e = strtok(name, ":"); i = 0;
	while(i < 6 && e != NULL) {
		mask[i] = 0xff;
		answer[i++] = ahextoi(e);
		e = strtok(NULL, ":");
	}

	return TRUE;
}

/*==============================================================================
| ip - stolen from config.y
|
|  this ... this is the most confusing piece of code i have ever seen in my
|  life.  i have no idea what it does, *exactly*  it... does a name lookup
|  or something odd, i can't quite figure out.  i certainly wish i had commented
|  it better when i wrote it, because i can't imagine what i was smoking when
|  i wrote it.  *shrug*
==============================================================================*/

static int
modifyip(u32 *num, char *name)
{
	u8	*tmp = (char *) num;
	char	*buff, *element;
	int	i=0;

	if((*num=searchbyname(name))!=0) {
		return TRUE;
	}

	buff = allocate(strlen(name) + 1);
	strcpy(buff, name);

	element = strtok(buff, ".");
		while(i < 4) {
			if(element == NULL) {

				fprintf(stderr, "warning: could not parse ip address %s\n",
					name);
				return FALSE;
			}
		tmp[i++] = atoi(element);
		element = strtok(NULL, ".");
	}

	return TRUE;
}

